src/pcap/pcap.go

```diff
@@ -137,3 +131,155 @@ func validateDevice(name string) (err error) {
+//
+// returns errNoMetadata if no metadata was found
+// returns errNoVcapID if no vcap-id was found in the metadata.
+func vcapIDFromOutgoingCtx(ctx context.Context) (*string, error) {
+	var vcap *string
+	var err error
+	if md, ok := metadata.FromOutgoingContext(ctx); ok {
+		vcap, err = getVcapFromMD(md)
+		if err == nil {
+			return vcap, nil
+		}
+	}
+
+	return nil, errNoMetadata
+}
+
+// vcapIDFromIncomingCtx finds the vcap-id from the context metadata, if available.
+//
+// returns errNoMetadata if no metadata was found
+// returns errNoVcapID if no vcap-id was found in the metadata.
+func vcapIDFromIncomingCtx(ctx context.Context) (*string, error) {
+	var vcap *string
+	var err error
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		vcap, err = getVcapFromMD(md)
+		if err == nil {
+			return vcap, nil
+		}
+	}
+	return nil, errNoMetadata
+}
+
+func getVcapFromMD(md metadata.MD) (*string, error) {
+	vcapReqIDs := md.Get(HeaderVcapID)
+
+	if len(vcapReqIDs) > 0 {
+		vcapID := vcapReqIDs[0]
+		return &vcapID, nil
+	}
+	return nil, errNoVcapID
+}
+
+// interfaceAddrs provides a list of all known network addresses.
+var interfaceAddrs = net.InterfaceAddrs
+
+// containsForbiddenRunes checks whether a given string contains
+// any character that is less than 32 or more than 126.
+//
+// See: https://www.lookuptables.com/text/ascii-table
+func containsForbiddenRunes(in string) bool {
+	for _, r := range in {
+		if r < 32 || r > 126 {
+			return true
+		}
+	}
+	return false
+}
+
+// patchFilter extends the given filter by excluding the filter generated
+// by generateApiFilter.
+func patchFilter(filter string) (string, error) {
+	apiFilter, err := generateAPIFilter()
+	if err != nil {
+		return "", err
+	}
+
+	filter = strings.TrimSpace(filter)
+
+	if filter == "" {
+		return fmt.Sprintf("not (%s)", apiFilter), nil
+	}
+
+	return fmt.Sprintf("not (%s) and (%s)", apiFilter, filter), nil
+}
+
+// generateApiFilter takes all IP addresses as returned by interfaceAddrs and
+// generates a filter for those IP addresses (loopback is excluded from the filter).
+// Note: the filter *matches* all of those IP addresses.
+func generateAPIFilter() (string, error) {
+	addrs, err := interfaceAddrs()
+	if err != nil {
+		return "", fmt.Errorf("unable to get IPs: %w", err)
+	}
+	if len(addrs) == 0 {
+		return "", fmt.Errorf("unable to determine ip addresses")
+	}
+
+	var ipFilters []string
+	for _, addr := range addrs {
+		ipNet, ok := addr.(*net.IPNet)
+		// check that:
+		// * ipNet is actually an IP address
+		// * it is not a loopback address
+		// * can be represented in either 4- or 16-bytes representation
+		if ok && !ipNet.IP.IsLoopback() {
+			v4 := ipNet.IP.To4() != nil
+			v6 := !v4 && ipNet.IP.To16() != nil
+
+			expression := "ip"
+			if !v4 && !v6 {
+				return "", fmt.Errorf("address %s is not IPv4 or v6", ipNet.IP.String())
+			}
+			if v6 {
+				expression = "ip6"
+			}
```

Comment:
This is confusing, please write it in a more idiomatic way, for example:
```suggestion
			var expression string
			switch {
			case v4:
				expression := "ip"
			case v6:
				expression := "ip6"
			default:
				return "", fmt.Errorf("address %s is not IPv4 or v6", ipNet.IP.String())
			}
```
With this you can replace the shorthands `v4` and `v6` with their full expressions again
```suggestion
			// Check whether the IP is v4 or v6. If both evaluate to true
			// v4 takes precedence.
			var expression string
			switch {
			case ipNet.IP.To4() != nil:
				expression := "ip"
			case ipNet.IP.To16() != nil:
				expression := "ip6"
			default:
				return "", fmt.Errorf("address %s is not IPv4 or v6", ipNet.IP.String())
			}
```

Reply:
Thanks. I keep forgetting that `switch` in Go is all-powerful.

Reply:
done